### ISSUE TITLE
Models with snapshotProcessor can't be added to maps

### ISSUE THREAD

**Issue by igaponov** [mobxjs/mobx-state-tree#1506]
**_Bug report_**

* [x] I've checked documentation and searched for existing issues
* [x] I've made sure my project is based on the latest MST version
* [x] Fork [this](https://codesandbox.io/s/y64pzxj01) code sandbox or another minimal reproduction. 

**Sandbox link or minimal reproduction code**
https://codesandbox.io/s/mobx-state-tree-todolist-i5idw

**Describe the expected behavior**
Maps should work with processed models without errors.

**Describe the observed behavior**
Adding a processed model throws the next exception:
```
[mobx-state-tree] Error while converting <Todo@<root>(id: 01)> to `Todo`:

    value of type Todo: <Todo@<root>(id: 01)> is not assignable to type: `Todo`, expected an instance of `Todo` or a snapshot like `snapshotProcessor({ id: identifier; title: string; finished: boolean? })` instead. (Note that a snapshot of the provided value is compatible with the targeted type)
```

**Comment by mweststrate** [comment]
I think you linked the wrong sandbox? It doesn't seem to contain either a map or a snapshot processor

**Comment by igaponov** [comment]
@mweststrate sorry, it didn't save my last changes. Now the link works.

**Comment by mbest** [comment]
I'm getting a similar problem:

> Error: [mobx-state-tree] Error while converting <WebFilterSettingsBase@<root>> to `(WebFilterSettingsBase | undefined)`:
    value of type WebFilterSettingsBase: <WebFilterSettingsBase@<root>> is not assignable to type: `(WebFilterSettingsBase | undefined)`, expected an instance of `(WebFilterSettingsBase | undefined)` or a snapshot like `(snapshotProcessor({ allowedList: string[]?; blockedList: string[]?; hasSettings: boolean; isWebsiteFilterEnabled: boolean; useAllowedListOnly: boolean; hasAnyWebsites: boolean; isBlockedBrowsersEnabled: boolean }) | undefined?)` instead. (Note that a snapshot of the provided value is compatible with the targeted type)

This is from simply trying to assign to a `types.maybe(WebFilterSettings)` using `value = WebFilterSettings.create(...)`. Below is the model definition.

```ts
const WebFilterSettingsBase = t.model("WebFilterSettingsBase", {
    allowedList: t.array(t.string),
    blockedList: t.array(t.string),
    hasSettings: t.boolean,
    isWebsiteFilterEnabled: t.boolean,
    useAllowedListOnly: t.boolean,
    hasAnyWebsites: t.boolean,
    isBlockedBrowsersEnabled: t.boolean,
}).actions(self => ({
    setIsWebsiteFilterEnabled: flow(function* setIsWebsiteFilterEnabled(newValue: boolean) {
        const settings = getParentOfType(self, ChildSettingsBase);
        yield setWebsiteFilterToggle(settings.child.userModelIds, newValue);
        self.isBlockedBrowsersEnabled = newValue;
    })
}));

export const WebFilterSettings = t.snapshotProcessor(WebFilterSettingsBase, {
    preProcessor(webFilterSettings: ApiWebFilterSettings) {
        return webFilterSettings as any;
    }
});
```

If you're wondering why I added the `snapshotProcessor`, it's because of the readonly array issue (https://github.com/mobxjs/mobx-state-tree/issues/605#issuecomment-618052315).

**Comment by mbest** [comment]
I tried this again after updating to 3.16, and the problem is still there.

**Comment by garrett-gottlieb** [comment]
@mweststrate any idea? I have another issue like this too.

**Comment by mweststrate** [comment]
No. Build reproductions folks if you want to report something. Closing this issue as the original problem was solved.

**Comment by garrett-gottlieb** [comment]
@mweststrate what was the solution?

**Comment by mweststrate** [comment]
Sorry, misread the second comment. I don't think there is a solution yet and looks like a bug at first sight.

**Comment by ashitikov** [comment]
Same problem. Please review #1581

**Comment by ashitikov** [comment]
Any progress on that?

**Comment by jamonholmgren** [comment]
@ashitikov I'm taking over as primary maintainer and plan to release #1581 shortly. Note that it will be in version 4.0.x -- if you can't upgrade to v4, I could perhaps do a patch release in 3.x.

**Comment by jamonholmgren** [comment]
We actually released #1581 in 5.0.0 just now:

https://github.com/mobxjs/mobx-state-tree/releases/tag/v5.0.0
